File: src/core/functions/circularShiftObfuscation.ts
// Obfuscation function: Circular shift
    function circularShiftObfuscation(input: Uint8Array, keyArray: Uint8Array, initor:any): Uint8Array {
        const obfuscated = new Uint8Array(input.length);

        for (let i = 0; i < input.length; i++) {
            obfuscated[i] = (input[i] << 1) | (input[i] >>> 7); // Circular left shift by 1 bit
        }

        return obfuscated;
    }

    // Deobfuscation function: Circular shift
    function de_circularShiftObfuscation(input: Uint8Array, keyArray: Uint8Array, initor:any): Uint8Array {
        const deobfuscated = new Uint8Array(input.length);

        for (let i = 0; i < input.length; i++) {
            deobfuscated[i] = (input[i] >>> 1) | (input[i] << 7); // Circular right shift by 1 bit
        }

        return deobfuscated;
    }

    let funPair = {
        obfuscation: circularShiftObfuscation,
        deobfuscation: de_circularShiftObfuscation
    }
    
    export default funPair;

File: android/wireguard_obfuscation/obfuscationFuns/CircularShiftObfuscation.java
package com.morph_vpn.plugin.wireguard_obfuscation.obfuscationFuns;
import java.util.Arrays;

public class CircularShiftObfuscation {
    public static byte[] circularShiftObfuscation(byte[] input, byte[] keyArray, Object initor) {
        byte[] obfuscated = new byte[input.length];

        for (int i = 0; i < input.length; i++) {
            int value = input[i] & 0xFF;
            obfuscated[i] = (byte) ((value << 1) | (value >>> 7)); // Circular left shift by 1 bit
        }

        return obfuscated;
    }

    public static byte[] de_circularShiftObfuscation(byte[] input, byte[] keyArray, Object initor) {
        byte[] deobfuscated = new byte[input.length];

        for (int i = 0; i < input.length; i++) {
            int value = input[i] & 0xFF;
            deobfuscated[i] = (byte) ((value >>> 1) | (value << 7)); // Circular right shift by 1 bit
        }

        return deobfuscated;
    }

    public static void main(String[] args) {
        byte[] input = new byte[]{/* initialize input byte array */};
        byte[] keyArray = new byte[]{/* initialize keyArray byte array */};
        Object initor = null;

        byte[] obfuscated = circularShiftObfuscation(input, keyArray, initor);
        byte[] deobfuscated = de_circularShiftObfuscation(obfuscated, keyArray, initor);

        System.out.println("Input: " + Arrays.toString(input));
        System.out.println("Obfuscated: " + Arrays.toString(obfuscated));
        System.out.println("Deobfuscated: " + Arrays.toString(deobfuscated));
    }
}
